### DISCUSSION TITLE
Use parts of matchStrings in autoReplaceStringTemplate

### DISCUSSION THREAD

**Discussion by ThomasSteinbach** [renovatebot/renovate#29068] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitlab; Renovate '37-full'

### Please tell us more about your question or problem

I have created a custom manager:

```json
{
    "customManagers": [
        {
            "customType": "regex",
            "fileMatch": [
                "^.*\\.ya?ml$"
            ],
            "matchStrings": [
                "originalImage=(?<depName>.*?)\\s.*: *(?<currentValue>[vV]?\\d+[\\.\\-\\w+]*)(@(?<currentDigest>sha256:[a-f0-9]+))?"
            ],
            "datasourceTemplate": "docker"
        }
    ]
}
```

With this manager I can define the actual dependency in a comment before the line with the value/version. Here an example:

```yaml
# originalImage=the-actual/image
version: 1.2.3

# originalImage=another/image
image: my-virtual/image:1.2.3
```

In the previous example Renovate successfully detects:
* the-actual/image:1.2.3
* another/image:1.2.3

To enable Renovate to pin digests, I have read that I have to tell Renovate how to replace the original match. This is by defining autoReplaceStringTemplate.

But how can I re-use required parts of the "matchStrings" in my "autoReplaceStringTemplate"? This is required, because I don't know the whole match aforehand.

I tried to use a custom named capture group named "customPrelude", which didn't work:

```json
{
    "customManagers": [
        {
            "customType": "regex",
            "fileMatch": [
                "^.*\\.ya?ml$"
            ],
            "matchStrings": [
                "originalImage=(?<depName>.*?)\\s(?<customPrelude>.*: *)(?<currentValue>[vV]?\\d+[\\.\\-\\w+]*)(@(?<currentDigest>sha256:[a-f0-9]+))?"
            ],
            "autoReplaceStringTemplate": "originalImage={{{depName}}}\n{{{customPrelude}}}:{{{newValue}}}{{#if newDigest}}@{{{newDigest}}}{{/if}}",
            "datasourceTemplate": "docker"
        }
    ]
}
```

So is this even possible?
If not, is there a different approach to enable Renovate to pin digests?

**Comment by rarkins** [comment]
Try using both depName and packageName

**Reply by ThomasSteinbach** [reply]
Ok. That seems to be a bit "abusive" to use "depName" for the whole unknown string.
But that worked  - however not like expected. It seems that Renovate will trim all whitespaces from "depName".

Given the following dependency...

```yaml
# bahnhubImage=dbv1-builds-docker-stage-local.myrepo.de/alpine
version: 1.0.0-dev
```

... and the updated manager configuration ...

```json
    "customManagers": [
        {
            "customType": "regex",
            "fileMatch": [
                "^.*\\.ya?ml$"
            ],
            "matchStrings": [
                "bahnhubImage=(?<depName>(?<packageName>.*?)\\s[\\w\\s\\:\\-\\/]*:\\s*)(?<currentValue>[vV]?\\d+[\\.\\-\\w+]*)(@(?<currentDigest>sha256:[a-f0-9]+))?"
            ],
            "autoReplaceStringTemplate": "bahnhubImage={{{depName}}}{{{newValue}}}{{#if newDigest}}@{{{newDigest}}}{{/if}}",
            "datasourceTemplate": "docker"
        }
    ]
```

The merge request generated by renovate will omit the whitespace between `version:` and the newValue:
<img width="982" alt="Bildschirmfoto 2024-05-15 um 14 42 05" src="https://github.com/renovatebot/renovate/assets/1683246/d3501c82-00c2-4719-9e45-bde5a6702a41">

When inserting the whitespace forcefully into...

```
"autoReplaceStringTemplate": "bahnhubImage={{{depName}}}\\t{{{newValue}}}{{#if newDigest}}@{{{newDigest}}}{{/if}}",
```

... will lead to an error creating the Branch. It seems that Renovate couldn't verify the original 'matchStrings' to the newly generated output.

Any idea of how to fix that (hopefully final issue)?

**Reply by rarkins** [reply]
I'm not sure. Do you have a way of reproducing this in a public repo?

**Reply by ThomasSteinbach** [reply]
@rarkins very good idea at this point. The example is very easy and the outcome easy to reproduce.

I just created this repository: https://github.com/ThomasSteinbach/renovate-29068-example/tree/original-issue

- It contains the fully working [renovate.json](https://github.com/ThomasSteinbach/renovate-29068-example/blob/original-issue/renovate.json) config file.
- It contains the [example-deps.yaml](https://github.com/ThomasSteinbach/renovate-29068-example/blob/original-issue/example-deps.yaml) with the two different dependency declarations the custom manager should be able to process.
- I activated the Renovate bot for this repo. (First time I did this on Github. Great experience 👏 )
- The outcome of the Renovate bot is the exact same behavior as on my server setup. It creates the pin-dependencies PR, with the missing blank between `version:` and `3.19.1...`
  👉 https://github.com/ThomasSteinbach/renovate-29068-example/pull/1/files#diff-03b37675607c8e8cd5fdab586bcfcedb2c6e57127bd031931ddd092e22b2274aR3

I spent hours with experiments, to make Renovate keeping the blank/space.
However I have a lack of understanding of how Renovate works under the hood and did not succeed.

Feel free to clone the repo and experiment with the code.

**Reply by ThomasSteinbach** [reply]
BTW: The behavior remains the same when updating an existing version with digest:

<img width="502" alt="Bildschirmfoto 2024-05-15 um 17 24 29" src="https://github.com/renovatebot/renovate/assets/1683246/cf029e06-f042-41e9-8d4a-f4368f6e7b7f">

**Reply by rarkins** [reply]
I took a look at the reproduction repo and unfortunately wasn't able to find a quick solution. I think you should check in case you can simplify it. It seems like you have made it intentionally difficult even. 

For example, 

```
# originalImage=alpine
version: 3.19.1
alsIgnore: thislinetoo
```

If you want it ignored then don't add the `# originalImage` line above it.

Next, this will be a lot easier if you do the digest pinning yourself and the rule is:
- If a digest is present, update it
- Otherwise, don't pin a digest

**Reply by ThomasSteinbach** [reply]
Is there any chance to see how Renovate is working under the hood and log/see what the autoReplace is doing?

I tried to simplify my custom manager and he fails to extract the updated dependency after autoreplace:

```
DEBUG: Could not extract example-deps.yaml (manager=regex) after autoreplace. Did the autoreplace make the file unparseable? (branch="renovate/pin-dependencies")
WARN: Error updating branch: update failure (branch="renovate/pin-dependencies")
```
Source: https://developer.mend.io/github/ThomasSteinbach/renovate-29068-example/-/job/1185ef91-04eb-4ae9-a882-2cb80ec668d1

However I can't see, what I am doing wrong until I can see what new string Renovate is creating after auto replace.

**Reply by rarkins** [reply]
You can see under the hood if you debug from source locally using e.g. Visual Studio Code

https://github.com/renovatebot/renovate/blob/main/docs/development/local-development.md

**Reply by ThomasSteinbach** [reply]
I think I have found the source of the unexpected behavior. It is possibly a **bug**.

The source of the problem is that Renovate trims all whitespaces from `depName` [at one point of the code (link)](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/process/fetch.ts#L29).

The problem is then two fold:
- If you use `depName` in `autoReplaceStringTemplate`, the resulting string does not match the original dependency anymore (because of the missing whitespace). This leads to the error I posted before (_...Could not extract ... after autoreplace._)
- If you manually insert the whitespace in the `autoReplaceStringTemplate`, then renovate fails on another point in the Code, where it doesn't trim the whitespace from `depName` [link to the code](https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/update/branch/auto-replace.ts#L73). You get a `depName mismatch` error:

```
DEBUG: depName mismatch (branch="renovate/pin-dependencies")
{
  "manager": "regex"
  "packageFile": "example-deps.yaml"
  "currentDepName": "version:"
  "newDepName": "version: "
}
```
Source: https://developer.mend.io/github/ThomasSteinbach/renovate-29068-example/-/job/6e190d8a-ad0c-47c0-9197-b231da47abc4

I have updated (and simplified) my example repo from above for this and created a git tag: https://github.com/ThomasSteinbach/renovate-29068-example/blob/bug-found/renovate.json#L20

@rarkins do you agree with me, that this is a bug?

**Reply by ThomasSteinbach** [reply]
[Update 2024-05-17 15:50] This post is obsolete as with the final solution posted below.

---

Eureka! 🥳

**I've found a working solution.**

Knowing the previously described behavior of Renovate, I've created two custom managers. They are both equal, despite the blank between the last colon before the version and the version itself.
👉 https://github.com/ThomasSteinbach/renovate-29068-example/blob/working-state/renovate.json

I also made the colon and blank (`: `) and not part of the `depName`. Knowing that those definetly exist (in the first manager), I could separately add them to the `autoReplaceStringTemplate`. Thus I have no trouble, that Renovate will trim the blank away.

Of course it would be better, if
- Renovate would use the same `depName` in the `autoReplaceStringTemplate` as discovered by the `matchStrings` regex.
- Renovate would allow custom handlebars in `matchStrings` that one could use in `autoReplaceStringTemplate`.

Nevertheless thank you @rarkins for looking into this. Your quick answers encouraged me to investigate further.

**Reply by rarkins** [reply]
Thanks for sharing your solution!

Allowing non-trimmed depNames is a hard decision to make. There's probably a few people who have regex which has mistakenly extracted a depName with unintended additional whitespace and changing this could _break_ them.

On the second matter, I don't fully understand your suggestion but if you could provide an example or perhaps even code for us to look at, we can definitely consider it.

**Reply by ThomasSteinbach** [reply]
[Update 2024-05-17 15:50] This post is obsolete as with the final solution posted below.

---

I am no JS/TS coder, so I try to describe my second suggestion in more detail:

The custom manager from my example can handle Docker dependencies, where the Docker registry / repository (`packageName`) is defined as meta information (comment) on one line and the Docker tag (`currentValue`) on the nextline, with arbitrary text in between. Here again two examples:

```yaml
# originalImage=mydockerregistry.net/myimage
image: 1234567890.dkr.ecr.eu-central-1.amazonaws.com/myimage:1.2.3-20200101@sha256:abcdef
```

```yaml
image: 1234567890.dkr.ecr.eu-central-1.amazonaws.com/myimage
# originalImage=mydockerregistry.net/myimage
version: 1.2.3-20200101@sha256:abcdef
```

In both examples `image:` and `version:` are randomly chosen and could be completely different.

That means, I don't know what is between `packageName` and `currentVersion`. But I need to know this string, to re-use it in my `autoReplaceStringTemplate`.

With your suggestion from above (https://github.com/renovatebot/renovate/discussions/29068#discussioncomment-9436571) I used `depName` for this unknown part. However that is a misuse of `depName` because this string could be totally garbage. Or at least from the second example, `depName` would just be `version: `, which is not the intended use of `depName`.

For this it would be great, If you could define custom named capture groups in `matchStrings`, you could then re-use in your `autoReplaceStringTemplate`.

Here a super simple example:

```
"matchStrings": [
  "originalImage=(?<depName>.*?)\\s(?<customPrelude>.*:\\s*)(?<currentValue>\\d+\\.\\d\\.\\d)(@(?<currentDigest>sha256:[a-f0-9]+))?"
  ],
"autoReplaceStringTemplate": "originalImage={{{depName}}}\n{{{customPrelude}}}:{{{newValue}}}{{#if newDigest}}@{{{newDigest}}}{{/if}}",
```

In this example `customPrelude` could be a custom named capture group introduced in `matchStrings`, which I can easily use in `autoReplaceStringTemplate`.

**Reply by ThomasSteinbach** [reply, marked as answer]
**I've found a way better solution than before.**

With this solution I ship around the "blank problem" and also don't need any custom named capture groupd.

For this solution I needed to understand, that **Renovate only replaces the last match, when using `recursive` match strategy**.
So in the last match I can focus on detecting the version and possible digest.
I didn't need to bother with whats between the version and the package name.

I have update my example with the latest tag: https://github.com/ThomasSteinbach/renovate-29068-example/blob/best-solution/renovate.json

The essence is, that in the last regex I only match the version. Then there is no need for a complex `autoReplaceStringTemplate`.

```
"matchStrings": [
  "originalImage=(?<depName>.*?)\\n[^\\n]+",
  "(?<currentValue>\\d+\\.\\d+\\.\\d+[\\w\\-\\.]*)(@(?<currentDigest>sha256:[a-f0-9]+))?"
],
"autoReplaceStringTemplate": "{{{newValue}}}{{#if newDigest}}@{{{newDigest}}}{{/if}}",
```
